File: apps/desktop/src-tauri/src/multiplayer.rs
//! Multiplayer: share a terminal tab with teammates, or watch someone else's.
//!
//! The heavy lifting (link secret, key derivation, frame encryption, relay
//! socket) lives in `termoso_core::live`; this module only ties it to the
//! session registry and turns relay events into Tauri events. The share
//! secret never reaches TypeScript except as the opaque link the host copies.

use std::collections::HashMap;
use std::sync::Mutex;

use serde::Serialize;
use tauri::{AppHandle, Emitter, Manager, Runtime};
use termoso_core::live::{HostShare, LiveEvent, LiveLink, ViewerJoin};
use termoso_core::terminal::TermSize;
use termoso_proto::live::LiveParticipant;
use tokio::sync::mpsc;
use uuid::Uuid;

use crate::account::api;
use crate::error::{DesktopError, Result};
use crate::sessions::Opened;
use crate::state::AppState;

pub const LIVE_EVENT: &str = "multiplayer";
/// `SessionInfo::protocol` of a tab that watches someone else's terminal.
pub const PROTOCOL: &str = "multiplayer";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum LiveRole {
    Host,
    Viewer,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Participant {
    pub user_id: Uuid,
    pub email: String,
    pub display_name: Option<String>,
    pub is_host: bool,
    pub can_write: bool,
    pub is_me: bool,
}

/// What the tab popover shows.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ShareInfo {
    /// Local session (tab) id.
    pub id: Uuid,
    pub live_id: Uuid,
    pub role: LiveRole,
    /// Host only: the `termoso://join/…` link to hand out.
    pub link: Option<String>,
    pub participants: Vec<Participant>,
    /// Viewer: whether our keystrokes reach the host terminal.
    pub can_write: bool,
}

/// Emitted as the `multiplayer` event.
#[derive(Debug, Clone, Serialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum LiveUiEvent {
    Participants {
        id: Uuid,
        participants: Vec<Participant>,
    },
    Control {
        id: Uuid,
        can_write: bool,
    },
    Resize {
        id: Uuid,
        cols: u16,
        rows: u16,
    },
    Title {
        id: Uuid,
        title: String,
    },
    Ended {
        id: Uuid,
        reason: String,
        message: String,
    },
}

struct Entry {
    live_id: Uuid,
    role: LiveRole,
    link: Option<String>,
    me: Uuid,
    participants: Vec<LiveParticipant>,
    can_write: bool,
    /// Host side only: dropping it disconnects the relay.
    share: Option<HostShare>,
}

impl Entry {
    fn info(&self, id: Uuid) -> ShareInfo {
        ShareInfo {
            id,
            live_id: self.live_id,
            role: self.role,
            link: self.link.clone(),
            participants: participants(&self.participants, self.me),
            can_write: self.can_write,
        }
    }
}

fn participants(list: &[LiveParticipant], me: Uuid) -> Vec<Participant> {
    let mut v: Vec<Participant> = list
        .iter()
        .map(|p| Participant {
            user_id: p.user_id,
            email: p.email.clone(),
            display_name: p.display_name.clone(),
            is_host: p.is_host,
            can_write: p.can_write,
            is_me: p.user_id == me,
        })
        .collect();
    v.sort_by_key(|p| (!p.is_host, !p.is_me, p.email.clone()));
    v
}

/// Live shares and views, keyed by local session id.
#[derive(Default)]
pub struct Multiplayer {
    entries: Mutex<HashMap<Uuid, Entry>>,
}

impl Multiplayer {
    pub fn info(&self, id: Uuid) -> Option<ShareInfo> {
        self.entries
            .lock()
            .expect("multiplayer poisoned")
            .get(&id)
            .map(|e| e.info(id))
    }

    /// Forget a session (tab closed / relay ended). A hosted share is stopped
    /// on the server so viewers hear about it right away.
    pub fn detach(&self, id: Uuid) {
        let entry = self
            .entries
            .lock()
            .expect("multiplayer poisoned")
            .remove(&id);
        if let Some(share) = entry.and_then(|e| e.share) {
            tokio::spawn(share.stop());
        }
    }

    fn take_share(&self, id: Uuid) -> Option<HostShare> {
        self.entries
            .lock()
            .expect("multiplayer poisoned")
            .get_mut(&id)
            .and_then(|e| e.share.take())
    }

    /// The shared terminal was resized by the host.
    pub fn resized(&self, id: Uuid, size: TermSize) {
        if let Some(share) = self
            .entries
            .lock()
            .expect("multiplayer poisoned")
            .get(&id)
            .and_then(|e| e.share.as_ref())
        {
            share.resized(size);
        }
    }
}

/// Start sharing a session. Fails when no account is signed in, when a team
/// disabled multiplayer, or when the tab is a viewer of someone else's share.
pub async fn start<R: Runtime>(app: &AppHandle<R>, id: Uuid) -> Result<ShareInfo> {
    let state = app.state::<AppState>();
    if let Some(existing) = state.multiplayer.info(id) {
        return match existing.role {
            LiveRole::Host => Ok(existing),
            LiveRole::Viewer => Err(DesktopError::invalid(
                "this tab is already a multiplayer view",
            )),
        };
    }
    let api = api(app).await?;
    let info = state.sessions.info(id)?;
    let term = state.sessions.terminal(id)?;
    let size = state.sessions.size(id)?;

    let (tx, rx) = mpsc::channel(64);
    let share = HostShare::start(api, term, size, info.title.clone(), tx).await?;
    let entry = Entry {
        live_id: share.session_id(),
        role: LiveRole::Host,
        link: Some(share.link().to_string()),
        me: share.user_id(),
        participants: Vec::new(),
        can_write: true,
        share: None,
    };
    let publisher = share.publisher();
    let result = entry.info(id);
    {
        let mut entries = state
            .multiplayer
            .entries
            .lock()
            .expect("multiplayer poisoned");
        entries.insert(
            id,
            Entry {
                share: Some(share),
                ..entry
            },
        );
    }
    if let Err(e) = state.sessions.set_tap(id, Some(publisher)) {
        // Tab vanished while we were talking to the server.
        state.multiplayer.detach(id);
        return Err(e);
    }
    tokio::spawn(forward(app.clone(), id, rx));
    Ok(result)
}

/// Stop sharing a session (host). Viewers get `Ended`; the tab keeps running.
pub async fn stop<R: Runtime>(app: &AppHandle<R>, id: Uuid) -> Result<()> {
    let state = app.state::<AppState>();
    let Some(share) = state.multiplayer.take_share(id) else {
        return Err(DesktopError::not_found("multiplayer session"));
    };
    let _ = state.sessions.set_tap(id, None);
    share.stop().await;
    Ok(())
}

/// Grant or revoke remote control for a viewer (host only).
pub fn set_control<R: Runtime>(
    app: &AppHandle<R>,
    id: Uuid,
    user_id: Uuid,
    enabled: bool,
) -> Result<()> {
    let state = app.state::<AppState>();
    let entries = state
        .multiplayer
        .entries
        .lock()
        .expect("multiplayer poisoned");
    let share = entries
        .get(&id)
        .and_then(|e| e.share.as_ref())
        .ok_or_else(|| DesktopError::not_found("multiplayer session"))?;
    share.set_control(user_id, enabled)?;
    Ok(())
}

/// Open someone else's terminal from a `termoso://join/…` link. Called by the
/// session registry as the `Live` open target; the tab then behaves like any
/// other session (output stream, resize, close).
pub(crate) async fn join<R: Runtime>(app: &AppHandle<R>, id: Uuid, link: &str) -> Result<Opened> {
    let state = app.state::<AppState>();
    let link = LiveLink::parse(link)?;
    let api = api(app).await?;
    if let Some(server) = &link.server {
        let ours = api.server_url();
        if server.host_str() != ours.host_str()
            || server.port_or_known_default() != ours.port_or_known_default()
        {
            return Err(DesktopError::invalid(format!(
                "this link belongs to {server}, but you are signed in to {ours}"
            )));
        }
    }
    let (tx, rx) = mpsc::channel(64);
    let ViewerJoin {
        term,
        events,
        user_id,
        participants,
    } = termoso_core::live::join(api, &link, tx).await?;
    let host = participants
        .iter()
        .find(|p| p.is_host)
        .map(|p| p.display_name.clone().unwrap_or_else(|| p.email.clone()))
        .unwrap_or_else(|| "Multiplayer".to_string());
    state
        .multiplayer
        .entries
        .lock()
        .expect("multiplayer poisoned")
        .insert(
            id,
            Entry {
                live_id: link.session_id,
                role: LiveRole::Viewer,
                link: None,
                me: user_id,
                participants,
                can_write: false,
                share: None,
            },
        );
    tokio::spawn(forward(app.clone(), id, rx));
    Ok(Opened {
        protocol: PROTOCOL,
        title: host.clone(),
        target: format!("Multiplayer · {host}"),
        host_id: None,
        vault_id: None,
        term,
        events,
        client: None,
        jumps: Vec::new(),
        startup: String::new(),
        color_scheme: None,
        shell: None,
    })
}

/// Turn relay events into Tauri events and keep the registry current.
async fn forward<R: Runtime>(app: AppHandle<R>, id: Uuid, mut rx: mpsc::Receiver<LiveEvent>) {
    let state = app.state::<AppState>();
    while let Some(ev) = rx.recv().await {
        let ui = {
            let mut entries = state
                .multiplayer
                .entries
                .lock()
                .expect("multiplayer poisoned");
            let Some(e) = entries.get_mut(&id) else { break };
            match ev {
                LiveEvent::Participants { participants } => {
                    e.participants = participants;
                    LiveUiEvent::Participants {
                        id,
                        participants: crate::multiplayer::participants(&e.participants, e.me),
                    }
                }
                LiveEvent::Control { can_write } => {
                    e.can_write = can_write;
                    LiveUiEvent::Control { id, can_write }
                }
                LiveEvent::Resize { cols, rows } => LiveUiEvent::Resize { id, cols, rows },
                LiveEvent::Title { title } => LiveUiEvent::Title { id, title },
                LiveEvent::Ended { reason, message } => {
                    let entry = entries.remove(&id);
                    drop(entries);
                    if entry.map(|e| e.role) == Some(LiveRole::Host) {
                        let _ = state.sessions.set_tap(id, None);
                    }
                    let _ = app.emit(
                        LIVE_EVENT,
                        LiveUiEvent::Ended {
                            id,
                            reason,
                            message,
                        },
                    );
                    break;
                }
            }
        };
        let _ = app.emit(LIVE_EVENT, ui);
    }
}

File: apps/desktop/src/terminal/multiplayer.ts
import { toast } from "@/components/Snackbar";
import * as ipc from "@/ipc/commands";
import {
  errorMessage,
  isDesktopError,
  type LiveEvent,
  type ShareInfo,
  type Uuid,
} from "@/ipc/types";
import { createStore, omit, useStore } from "@/lib/store";
import { setPaneFixedSize, terminalStore, writeSystemLine } from "./store";
import { tr } from "@/i18n";

/** Shared (host) and watched (viewer) panes, by pane id. */
export interface MultiplayerState {
  shares: Record<Uuid, ShareInfo>;
  /** Viewer pane that was just handed remote control — shows the "start typing" hint. */
  controlHint: Uuid | null;
}

export const multiplayerStore = createStore<MultiplayerState>({ shares: {}, controlHint: null });

export const useMultiplayer = <S>(selector: (s: MultiplayerState) => S) =>
  useStore(multiplayerStore, selector);

export const useShare = (paneId: Uuid | undefined) =>
  useMultiplayer((s) => (paneId ? (s.shares[paneId] ?? null) : null));

const update = (fn: (s: MultiplayerState) => MultiplayerState) =>
  multiplayerStore.set(fn(multiplayerStore.get()));

function put(share: ShareInfo) {
  update((s) => ({ ...s, shares: { ...s.shares, [share.id]: share } }));
}

function forget(paneId: Uuid) {
  update((s) => ({
    ...s,
    shares: omit(s.shares, paneId),
    controlHint: s.controlHint === paneId ? null : s.controlHint,
  }));
}

/** Is the multiplayer prohibited by one of the user's teams (server-side flag)? */
export const isMultiplayerDisabled = (e: unknown) =>
  isDesktopError(e) && e.kind === "multiplayer_disabled";

/** Not signed in to any Termoso account — sharing needs the relay. */
export const isNotSignedIn = (e: unknown) =>
  isDesktopError(e) && e.kind === "invalid" && /not signed in/i.test(e.message);

/** Start sharing a pane; resolves with the link to hand out. Rethrows so the UI can explain. */
export async function startShare(paneId: Uuid): Promise<ShareInfo> {
  const info = await ipc.multiplayerStart(paneId);
  put(info);
  return info;
}

export async function stopShare(paneId: Uuid) {
  try {
    await ipc.multiplayerStop(paneId);
  } finally {
    forget(paneId);
  }
}

export async function setControl(paneId: Uuid, userId: Uuid, enabled: boolean) {
  await ipc.multiplayerSetControl(paneId, userId, enabled);
  const who = multiplayerStore.get().shares[paneId]?.participants.find((p) => p.userId === userId);
  const name = who?.displayName ?? who?.email ?? "Participant";
  toast(
    enabled
      ? tr("{name} has remote control", { name })
      : tr("Remote control taken back from {name}", { name }),
    "info",
  );
  update((s) => {
    const share = s.shares[paneId];
    if (!share) return s;
    return {
      ...s,
      shares: {
        ...s.shares,
        [paneId]: {
          ...share,
          participants: share.participants.map((p) =>
            p.userId === userId ? { ...p, canWrite: enabled } : p,
          ),
        },
      },
    };
  });
}

export function dismissControlHint() {
  update((s) => (s.controlHint ? { ...s, controlHint: null } : s));
}

/** A viewer pane finished joining: pull the initial participant list. */
export async function viewerJoined(paneId: Uuid) {
  try {
    const info = await ipc.multiplayerInfo(paneId);
    if (info) {
      put(info);
      const host = info.participants.find((p) => p.isHost);
      const who = host?.displayName ?? host?.email ?? tr("the host");
      writeSystemLine(
        paneId,
        info.canWrite
          ? tr("Multiplayer: you are watching {who}'s terminal", { who })
          : tr("Multiplayer: you are watching {who}'s terminal (view only)", { who }),
      );
    }
  } catch (e) {
    toast(errorMessage(e), "error");
  }
}

function onLiveEvent(ev: LiveEvent) {
  const s = multiplayerStore.get();
  const share = s.shares[ev.id];
  switch (ev.type) {
    case "participants":
      if (share) put({ ...share, participants: ev.participants });
      break;
    case "control":
      if (share) put({ ...share, canWrite: ev.canWrite });
      if (share?.role === "viewer") {
        update((st) => ({ ...st, controlHint: ev.canWrite ? ev.id : null }));
        if (!ev.canWrite) writeSystemLine(ev.id, tr("Multiplayer: remote control revoked"));
      }
      break;
    case "resize":
      if (terminalStore.get().panes[ev.id]?.target.kind === "live") {
        setPaneFixedSize(ev.id, { cols: ev.cols, rows: ev.rows });
      }
      break;
    case "title":
      break;
    case "ended": {
      const role = share?.role;
      forget(ev.id);
      if (role === "host") {
        const pane = terminalStore.get().panes[ev.id];
        if (ev.reason !== "stopped") {
          toast(`Multiplayer for ${pane?.title ?? "terminal"} ended: ${ev.message}`, "warning");
        }
      } else if (role === "viewer") {
        writeSystemLine(ev.id, tr("Multiplayer: {message}", { message: ev.message }));
      }
      break;
    }
  }
}

let started = false;
/** Subscribe to Rust multiplayer events once for the app lifetime. */
export function startMultiplayerEvents() {
  if (started) return;
  started = true;
  void ipc.onLiveEvent(onLiveEvent);
  terminalStore.subscribe(() => {
    const panes = terminalStore.get().panes;
    for (const id of Object.keys(multiplayerStore.get().shares)) {
      if (!panes[id]) forget(id);
    }
  });
}
